sdk/modelsrepository/azure-iot-modelsrepository/src/main/java/com/azure/iot/modelsrepository/DtmiConventions.java

```diff
@@ -0,0 +1,122 @@
+     * The first digit may not be zero, so there is no ambiguity regarding whether version 1 matches version 01 since the latter is invalid.
+     */
+    private static final Pattern VALID_DTMI_PATTERN = Pattern.compile("^dtmi:[A-Za-z](?:[A-Za-z0-9_]*[A-Za-z0-9])?(?::[A-Za-z](?:[A-Za-z0-9_]*[A-Za-z0-9])?)*;[1-9][0-9]{0,8}$");
+
+    /**
+     * Indicates whether a given string DTMI value is well-formed.
+     *
+     * @param dtmi DigitalTwin Model Id
+     * @return True if dtmi has a valid format. False otherwise.
+     */
+    public static boolean isValidDtmi(String dtmi) {
+        if (dtmi == null || dtmi.isEmpty()) {
+            return false;
+        }
+
+        return VALID_DTMI_PATTERN.matcher(dtmi).find();
+    }
+
+    /**
+     * Generates the model URI.
+     *
+     * @param dtmi DigitalTwin Model Id.
+     * @param repositoryUri The repository uri
+     * @param expanded Is model from precomputed values
+     * @return The model uri
+     */
+    public static URI getModelUri(String dtmi, URI repositoryUri, boolean expanded) {
+        String dtmiPath = dtmiToPath(dtmi);
+
+        if (expanded) {
+            dtmiPath = dtmiPath.replace(ModelsRepositoryConstants.JSON_EXTENSION, ModelsRepositoryConstants.JSON_EXPANDED_EXTENSION);
+        }
+
+        UrlBuilder urlBuilder = new UrlBuilder();
+        urlBuilder.setHost(repositoryUri.getHost());
+        urlBuilder.setScheme(repositoryUri.getScheme());
+        urlBuilder.setPath(repositoryUri.getPath());
+        urlBuilder.setQuery(repositoryUri.getQuery());
+
+        if (repositoryUri.getPort() > 0) {
+            urlBuilder.setPort(repositoryUri.getPort());
+        }
+
+        String path = urlBuilder.getPath();
+
+        if (path != null && !path.endsWith("/")) {
+            urlBuilder.setPath(path + "/");
+        }
+
+        if (urlBuilder.getPath() == null) {
+            urlBuilder.setPath(dtmiPath);
+        } else {
+            urlBuilder.setPath(urlBuilder.getPath() + dtmiPath);
+        }
+
+        try {
+            return new URI(urlBuilder.toString());
+        } catch (Exception e) {
+            // No exceptions will be thrown as the input is a valid URI format.
+            return null;
+        }
+    }
+
+    /**
+     * Converts a string to {@link URI}
+     *
+     * @param uri String format of the path
+     * @return {@link URI} representation of the path/uri   .
+     */
+    public static URI convertToUri(String uri) {
+        try {
+            return new URI(uri);
+        } catch (URISyntaxException ex) {
+            Path path = Paths.get(uri).normalize();
+            return new File(path.toAbsolutePath().toString()).toURI();
+        }
+    }
+
+    static String dtmiToPath(String dtmi) {
+        if (!isValidDtmi(dtmi)) {
+            throw new IllegalArgumentException(String.format(StandardStrings.INVALID_DTMI_FORMAT_S, dtmi));
+        }
+
+        return dtmi
+            .toLowerCase(Locale.getDefault())
```

Comment:
Is the DTMI going to have different handling per locale or is it going to be ASCII based?

Reply:
it will be ASCII based.